conans/client/command.py

```diff
@@ -997,44 +995,52 @@ def search(self, *args):
+            remote_all = None
+        all_remotes_search = (remote_all is None and args.remote == "all")
+
+        out_kwargs = {'all_remotes_search': all_remotes_search, 'cwd': os.path.abspath(os.getcwd()),
+                      'out': self._user_io.out, 'raw': args.raw}
 
         try:
             if reference:
+                out_kwargs.update({'f': 'search_packages',
+                                   'packages_query': args.query,
+                                   'outdated': args.outdated,
+                                   'query': reference})
 
                 info = self._conan.search_packages(reference.full_repr(), query=args.query,
                                                    remote_name=args.remote,
                                                    outdated=args.outdated)
-                # search is done for one reference
-                self._outputer.print_search_packages(info["results"], reference, args.query,
-                                                     args.table, outdated=args.outdated)
+                OutputerFormats.get('cli').out(info=info, **out_kwargs)
             else:
+                out_kwargs.update({'f': 'search_recipes',
+                                   'query': args.pattern_or_reference})
+
                 if args.table:
                     raise ConanException("'--table' argument can only be used with a reference")
-                elif args.outdated:
+                if args.outdated:
                     raise ConanException("'--outdated' argument can only be used with a reference")
 
                 self._check_query_parameter_and_get_reference(args.pattern_or_reference, args.query)
 
                 info = self._conan.search_recipes(args.pattern_or_reference,
                                                   remote_name=args.remote,
                                                   case_sensitive=args.case_sensitive)
-                # Deprecate 2.0: Dirty check if search is done for all remotes or for remote "all"
-                try:
-                    remote_all = self._conan.get_remote_by_name("all")
-                except NoRemoteAvailable:
-                    remote_all = None
-                all_remotes_search = (remote_all is None and args.remote == "all")
-
-                self._outputer.print_search_references(info["results"], args.pattern_or_reference,
-                                                       args.raw, all_remotes_search)
+                OutputerFormats.get('cli').search_recipes(info=info, **out_kwargs)
         except ConanException as exc:
             info = exc.info
             raise
         finally:
-            if args.json and info:
-                self._outputer.json_output(info, args.json, cwd)
+            if args.json:
+                filepath = os.path.normpath(os.path.join(os.getcwd(), args.json))
+                OutputerFormats.get('json').out(output_filepath=filepath, info=info, **out_kwargs)
```

Comment:
I think passing the action as an argument in ``out_kwargs`` and then calling ``.out()`` is overengineered, and unnecessarily hiding things.

Reply:
I totally agree with you, but here we can have `search_recipes` or `search_packages`, so we can have the action in the `**kwargs` or put here (and two lines after) an `if` clause.